pkg/runtime/bulk_subscriber.go

```diff
@@ -0,0 +1,503 @@
+		Metadata: route.metadata,
+	}
+
+	bulkHandler := func(ctx context.Context, msg *pubsub.BulkMessage) ([]pubsub.BulkSubscribeResponseEntry, error) {
+		if msg.Metadata == nil {
+			msg.Metadata = make(map[string]string, 1)
+		}
+
+		msg.Metadata[pubsubName] = psName
+
+		rawPayload, err := contribMetadata.IsRawPayload(route.metadata)
+		if err != nil {
+			log.Errorf("error deserializing pubsub metadata: %s", err)
+			if dlqErr := a.sendBulkToDLQIfConfigured(ctx, psName, msg, route, nil, nil); dlqErr != nil {
+				return nil, err
+			}
+			return nil, nil
+		}
+		routePathBulkMessageMap := make(map[string]pubsubBulkSubscribedMessage)
+		bulkResponses := make([]pubsub.BulkSubscribeResponseEntry, len(msg.Entries))
+		entryIDIndexMap := make(map[string]int)
+		hasAnyError := false
+		for i, message := range msg.Entries {
+			if entryIDErr := validateEntryID(message.EntryID, i); entryIDErr != nil {
+				bulkResponses[i].Error = entryIDErr
+				hasAnyError = true
+				continue
+			}
+			entryIDIndexMap[message.EntryID] = i
+			if rawPayload {
+				rPath, routeErr := a.getRouteIfProcessable(ctx, route, &(msg.Entries[i]), i, &bulkResponses, string(message.Event), psName, topic)
+				if routeErr != nil {
+					hasAnyError = true
+					continue
+				}
+				if rPath == "" {
+					continue
+				}
+				dataB64 := base64.StdEncoding.EncodeToString(message.Event)
+				var contenttype string
+				if message.ContentType != "" {
+					contenttype = message.ContentType
+				} else {
+					contenttype = "application/octet-stream"
+				}
+				populateBulkSubcribedMessage(&(msg.Entries[i]), dataB64, &routePathBulkMessageMap, rPath, i, msg, false, psName, contenttype)
+			} else {
+				var cloudEvent map[string]interface{}
+				err = json.Unmarshal(message.Event, &cloudEvent)
+				if err != nil {
+					log.Errorf("error deserializing one of the messages in bulk cloud event in pubsub %s and topic %s: %s", psName, msg.Topic, err)
+					bulkResponses[i].Error = err
+					bulkResponses[i].EntryID = message.EntryID
+					hasAnyError = true
+					continue
+				}
+				if pubsub.HasExpired(cloudEvent) {
+					log.Warnf("dropping expired pub/sub event %v as of %v", cloudEvent[pubsub.IDField], cloudEvent[pubsub.ExpirationField])
+					diag.DefaultComponentMonitoring.BulkPubsubIngressEvent(ctx, pubsubName, strings.ToLower(string(pubsub.Drop)), topic, 0)
+					if route.deadLetterTopic != "" {
+						_ = a.sendToDeadLetter(psName, &pubsub.NewMessage{
+							Data:        message.Event,
+							Topic:       topic,
+							Metadata:    message.Metadata,
+							ContentType: &message.ContentType,
+						}, route.deadLetterTopic)
+					}
+					bulkResponses[i].EntryID = message.EntryID
+					bulkResponses[i].Error = nil
+					continue
+				}
+				rPath, routeErr := a.getRouteIfProcessable(ctx, route, &(msg.Entries[i]), i, &bulkResponses, cloudEvent, psName, topic)
+				if routeErr != nil {
+					hasAnyError = true
+					continue
+				}
+				if rPath == "" {
+					continue
```

Comment:
What does it mean when route is nil, the message cannot be sent to app so why not set the bulk response to component as error there it self?

Reply:
Actually, this route is calculated against each message/entry inside a bulk message; so we just need to catch appropriate error for this message only (if any), and then continue for other messages/entries in this bulk message.